src/librustc_mir/borrow_check.rs

```diff
@@ -1600,7 +1611,20 @@ impl<'c, 'b, 'a: 'b+'c, 'gcx, 'tcx: 'a> MirBorrowckCtxt<'c, 'b, 'a, 'gcx, 'tcx>
                 },
                 ty::TyArray(ty, _) | ty::TySlice(ty) => {
                     self.describe_field_from_ty(&ty, field_index)
-                }
+                },
+                ty::TyClosure(closure_def_id, _) => {
+                    // Convert the def-id into a node-id. node-ids are only valid for
+                    // the local code in the current crate, so this returns an `Option` in case
+                    // the closure comes from another crate. But in that case we wouldn't
+                    // be borrowck'ing it, so we can just unwrap:
+                    let node_id = self.tcx.hir.as_local_node_id(closure_def_id).unwrap();
+                    let local_def = self.tcx.with_freevars(node_id, |fv| fv[field_index].def);
+
+                    match local_def {
+                        Def::Local(local_node_id) => self.tcx.hir.name(local_node_id).to_string(),
```

Comment:
I saw that [the code in `ppaux`](https://github.com/rust-lang/rust/blob/dbeb5bf890211608bff5d28c3801c1a217a78d81/src/librustc/util/ppaux.rs#L1063) does `self.tcx.hir.name(local_def.var_id())` -- maybe we should do that. I think in particular this could be a `Def::Upvar` for nested closures. Maybe try this example:

```rust
// Field from upvar
{
    let mut x = 0;
       || { || {
            let y = &mut x;
            &mut x; //[ast]~ ERROR cannot borrow `**x` as mutable more than once at a time
                    //[mir]~^ ERROR cannot borrow `**x` as mutable more than once at a time (Ast)
                    //[mir]~| ERROR cannot borrow `(**x)` as mutable more than once at a time (Mir)
            *y = 1;
        } };
}
```